packages/cli/lib/make-index.js

```diff
@@ -0,0 +1,109 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+const util = require('util');
+const glob = util.promisify(require('glob'));
+const path = require('path');
+const fs = require('fs');
+const format = require('./format');
+const debug = require('./debug')('make-index');
+
+/**
+ * A utility function to make index files to export artifacts from a folder
+ * for easier imports
+ *
+ * @param {string} dir The directory for which to generate index files.
+ * @param {object} options
+ * options.nested: boolean = true
+ * options.prefix: string | string[] = ''
+ * options.commentHeader:string = SEE BELOW
+ */
+module.exports = async function(dir, options = {}) {
+  // Only exporting typescript files.
+  // TODO: @virkt25 -- May want this to be configurable in the future?
+  const EXTENSION = '.ts';
+
+  // Default Options for Making an exports index file
+  const defaultOptions = {
+    nested: true,
+    prefix: '',
+    commentHeader: `This is an auto-generated file. DO NOT EDIT.`,
+    commentFooter: `This is an auto-generated file. DO NOT EDIT.`,
+  };
+
+  // Merge User Options with defaultOptions
+  options = Object.assign(defaultOptions, options);
+
+  // If prefix is array, convert to a string that can be used in a glob
+  options.prefix = Array.isArray(options.prefix)
+    ? `@(${options.prefix.join('|')})`
+    : options.prefix;
+
+  // Glob pattern for search
+  const globPattern = `*${options.prefix}${EXTENSION}`;
+
+  debug(`dir => ${dir}`);
+  debug(`options => ${JSON.stringify(options)}`);
+  debug(`globPattern =>: ${globPattern}`);
+
+  // Glob Options object for running glob search
+  const globOptions = {root: dir, matchBase: options.nested};
+
+  // Run a Glob Search
+  const files = await glob(globPattern, globOptions);
+  const lines = [];
+
+  // Add comment header as first line
+  if (options.commentHeader) {
+    lines.push(normalizeComment(options.commentHeader));
+  }
+
+  // For each file discovers by the glob pattern, generate an export statement
+  files.forEach(file => {
+    const relPath = `export * from './${path
+      .relative(dir, file)
+      .slice(0, -3)}';`;
+    lines.push(relPath);
+  });
+
+  // Add comment footer if present
+  if (options.commentFooter) {
+    lines.push('\n');
+    lines.push(normalizeComment(options.commentFooter));
+  }
+
+  // Turn array into a string and format it using prettier formatter
+  const code = format(lines.join('\n'));
+
+  debug('===== code =====');
+  debug(code);
+  debug('===== code =====');
+
+  const outFile = `${dir}/index.ts`;
+  debug(`outFile => ${outFile}`);
+
+  const writeFileAsync = util.promisify(fs.writeFile);
+  await writeFileAsync(outFile, code);
```

Comment:
This will break incremental builds, because every run of `make-index` is going to change the modification time of `index.ts` to the current date, even in cases where the file content was not changed at all.

Proposed solution:

 1. Read the current file content and if the old content is the same as the new content, then don't write anything. I believe this is what prettier is doing too.
 2. Set `mtime` of the written file to match the latest `mtime` of files we were generating index from. I.e. for each file in `dir` (excluding `index.ts`), obtain `mtime` value and then pick the maximum value of all mtimes.

I think the second item is not enough on its own, because file watchers like `tsc --watch`, fs-events, etc. listen for write events and ignore `mtime`.

The first item may is most likely good enough, because typically the `index.ts` file would be updated shortly after the last non-index file was added/updated.

Reply:
I am ok to leave these two improvements out of scope of this initial pull request as long as there is a follow-up issue to ensure we fix this problem as part of the DP3 Boot epic.